packages/dashboards/addon/services/dashboard-data.js

```diff
@@ -36,54 +39,89 @@ export default Service.extend({
   /**
    * @method fetchDataForDashboard
    * @param {Object} dashboard - dashboard model
-   * @returns {Promise} - Promise that resolves to a dashboard object with widget data
+   * @returns {Promise} - Promise that resolves to a hash of widget id to data promise array
+   */
+  async fetchDataForDashboard(dashboard) {
+    const widgets = await dashboard.widgets;
+    const layout = get(dashboard, 'presentation.layout');
+
+    return this.fetchDataForWidgets(dashboard.id, widgets, layout, [], get(this, 'widgetOptions'));
+  },
+
+  /**
+   * @method cancelFetchDataForDashboard - cancels all running or enqueued fetch Task Instances
    */
-  fetchDataForDashboard(dashboard) {
-    return dashboard
-      .get('widgets')
-      .then(widgets => this.fetchDataForWidgets(dashboard.id, widgets, [], get(this, 'widgetOptions')));
+  cancelFetchDataForDashboard() {
+    this._fetchTask.cancelAll();
   },
 
   /**
    * @method fetchDataForWidgets
+   * @param {Number} dashboardId
    * @param {Array} widgets - list of widget models with requests to fetch
+   * @param {Array} layout - dashboard layout
    * @param {Array} decorators - array of functions to modify each request
    * @param {Object} options - options for web service fetch
    * @returns {Object} hash of widget id to data promise array
    */
-  fetchDataForWidgets(dashboardId, widgets = [], decorators = [], options = {}) {
-    const result = {};
+  fetchDataForWidgets(dashboardId, widgets = [], layout = [], decorators = [], options = {}) {
     const uuid = v1();
 
-    // Construct hash of widget id to data
-    widgets.forEach(widget => {
-      const widgetId = get(widget, 'id');
-      const dashboard = get(widget, 'dashboard');
-      const widgetDataPromises = get(widget, 'requests').map(request => {
-        //construct custom header for each widget with uuid
-        options.customHeaders = {
-          uiView: `dashboard.${dashboardId}.${uuid}.${widgetId}`
-        };
+    // sort widgets by order in layout
+    const sortedWidgets = arr(layout)
+      .sortBy('row', 'column')
+      .map(layoutItem => widgets.find(widget => widget.id == layoutItem.widgetId))
+      .filter(widget => widget);
+
+    // For each widget, concurrently execute a task that will fetch all widget's requests
+    return sortedWidgets.reduce(
+      (dataByWidget, widget) => ({
+        [widget.id]: DS.PromiseArray.create({
```

Comment:
```js
dataByWidget[widget.id] = ...
return dataByWidget;
```

Reply:
for readability?

Reply:
It would be nice if we can get away from DS.PromiseArray. Looks like we have a mix of usage:

DS.PromiseArray:
https://github.com/yahoo/navi/blob/271bc51d48620c041367492919b823450b0716e1/packages/dashboards/addon/components/navi-widget.js#L62

Native Promise
https://github.com/yahoo/navi/blob/271bc51d48620c041367492919b823450b0716e1/packages/dashboards/addon/templates/components/navi-widget.hbs#L48

Reply:
> for readability?

Yea seems simpler to read than creating a new object